README.md
# butterfly

<div align=center><img src="https://github.com/meetbill/butterfly/blob/master/images/butterfly.png" width="350"/></div>

```
    __          __  __            ______
   / /_  __  __/ /_/ /____  _____/ __/ /_  __
  / __ \/ / / / __/ __/ _ \/ ___/ /_/ / / / /
 / /_/ / /_/ / /_/ /_/  __/ /  / __/ / /_/ /
/_.___/\__,_/\__/\__/\___/_/  /_/ /_/\__, /
                                    /____/
```

<!-- vim-markdown-toc GFM -->

* [1 简介](#1-简介)
    * [1.1 介绍](#11-介绍)
    * [1.2 环境](#12-环境)
    * [1.3 特性](#13-特性)
* [2 五分钟 Butterfly 体验指南](#2-五分钟-butterfly-体验指南)
    * [2.1 五分钟体验之部署 Butterfly（预计 1 分钟）](#21-五分钟体验之部署-butterfly预计-1-分钟)
    * [2.2 五分钟体验之编写 handler （预计 3 分钟）](#22-五分钟体验之编写-handler-预计-3-分钟)
    * [2.3 五分钟体验之调试 handler （预计 1 分钟）](#23-五分钟体验之调试-handler-预计-1-分钟)
* [3 设计蓝图](#3-设计蓝图)
    * [3.1 单实例内部请求流](#31-单实例内部请求流)
    * [3.2 多实例间消息队列通信](#32-多实例间消息队列通信)
* [4 了解更多](#4-了解更多)
    * [4.1 传送门](#41-传送门)
    * [4.2 自带 app 介绍](#42-自带-app-介绍)
    * [4.3 报告错误](#43-报告错误)
* [5 贡献](#5-贡献)
* [6 版本信息](#6-版本信息)
* [7 参加步骤](#7-参加步骤)

<!-- vim-markdown-toc -->

# 1 简介
## 1.1 介绍

Butterfly（轻量化 WSGI web 应用程序框架）如同蝴蝶一样，小而美，简单可依赖。目的是让入门变得简单快捷，能够扩展到复杂的应用程序。

> * <1> 接入 --- 【接入】函数快速适配：普通 Python 函数可快速升级为 butterfly handler
> * <2> 异步 --- 【百川】任务异步处理：任一 butterfly handler 皆可消费，异步处理 MQ 中消息
> * <3> 定时 --- 【如期】任务定时调度：设置 butterfly handler 执行时间，定期执行 handler
> * <4> 编排 --- 【星桥】任务依赖编排：设置 butterfly handler 依赖顺序，有序编排 handler

## 1.2 环境

```
env:Python 2.7
```
## 1.3 特性

> * 快速开发
>   * (1) 无需配置路由：根据 handlers package 下目录结构自动加载路由（目前不支持动态路由）
>   * (2) 参数保持一致：Handler 的参数列表与 HTTP 请求参数保持一致，HTTP 接口所需参数一目了然
>   * (3) 自动参数检查：自动对 HTTP 请求参数进行参数检查
>   * (4) 简易调试模式：简易方便的 DEBUG
>   * (5) 引擎之状态机：具有可复用性的状态处理
>   * (6) 引擎之工作流：长流程分步执行，可在运行关键点处进行 checkpoint，可生成 dot 流程图
>   * (7) 对象关系映射：自带 ORM
>   * (8) 定时任务调度：支持定时执行某些方法
> * 方便运维
>   * (1) 请求完整追溯：请求的响应 Header 中包含请求的 reqid（会记录在日志中）, 便于进行 trace
>   * (2) 自定义响应头：Header 函数中很方便自定义 HTTP header, 如增加固定的接口版本号
>   * (3) 代码耗时打点：通过代码打点可以准确获代码执行耗时
> * 容易扩展
>   * (1) 消息队列通信：启动时开启百川配置即成为一个消费者，以拉模式消费由其他实例发布的消息

# 2 五分钟 Butterfly 体验指南

> 流程简述
```
(1) 下载 butterfly 包
(2) 编写 handler
   开发人员主要就是编写 {butterfly_project}/handlers/{app}/{handler}
   * {app} 是 Python package, 即 Linux 系统上的 目录
   * {handler} 是 Python function
(3) 测试 handler
   python test_handler.py /{app}/{handler} param1 param2 ...
(4) 启动 butterfly 提供 web 服务
   curl -v "http://{IP}:{PORT}/{app}/{handler}?{param1}=value1&..."
```

## 2.1 五分钟体验之部署 Butterfly（预计 1 分钟）
> 部署
```
$ wget https://github.com/meetbill/butterfly/archive/master.zip -O butterfly.zip
$ unzip butterfly.zip
$ cd butterfly-master/butterfly
```
> 配置端口 --- 默认 8585 端口，若无需修改可进入下一项启动
```
conf/config.py
```
> 启动
```
$ bash run.sh start
```
> 访问
```
$ curl "http://127.0.0.1:8585/demo_api/ping"
{"stat": "OK", "randstr": "..."}

$ curl "http://127.0.0.1:8585/demo_api/hello?str_info=meetbill"
{"stat": "OK", "str_info": "meetbill"}
```
##  2.2 五分钟体验之编写 handler （预计 3 分钟）

> 创建 app （handlers 目录下的子目录均为 app）, 直接拷贝个现有的 app 即可
```
$ cp -rf handlers/demo_api handlers/test_app
```
> 新增 handler (app 目录下的`__init__.py` 中编写 handler 函数）
```
$ vim handlers/test_app/__init__.py

新增如下内容：

# ------------------------------ handler
@funcattr.api
def test_handler(req, info):
    return retstat.OK, {"data": info}
```
> 重启服务
```
$ bash run.sh restart
```

> 访问
```
$ curl "http://127.0.0.1:8585/test_app/test_handler?info=helloworld"
{"stat": "OK", "data": "helloworld"}
```

##  2.3 五分钟体验之调试 handler （预计 1 分钟）

假如编写的 handler 不符合预期的时候，可以通过 test_handler.py 进行调试

> 调试刚才编写的 test_handler
```
$ python test_handler.py /test_app/test_handler helloworld
... 此处会输出彩色的  DEBUG 信息
Source path:... test_handler.py
>>>>>|19:03:53.293076 4694912448-MainThread call        66             def main():
------19:03:53.294108 4694912448-MainThread line        67                 return func(*args)
    Source path:... /Users/meetbill/butterfly-master/butterfly/handlers/test_app/__init__.py
    Starting var:.. req = <xlib.httpgateway.Request object at 0x10b4148d0>
    Starting var:.. info = 'helloworld'
    >>>>>|19:03:53.294402 4694912448-MainThread call        49 def test_handler(req, info):
    ------19:03:53.294864 4694912448-MainThread line        50     return retstat.OK, {"data": info}
    |<<<<<19:03:53.294989 4694912448-MainThread return      50     return retstat.OK, {"data": info}
    Return value:.. ('OK', {'data': 'helloworld'})
Source path:... test_handler.py
|<<<<<19:03:53.295114 4694912448-MainThread return      67                 return func(*args)
Return value:.. ('OK', {'data': 'helloworld'})
Elapsed time: 00:00:00.002140
=============================================================
('OK', {'data': 'helloworld'})
=============================================================
```
运行中产生的 logging 日志会记录在 `logs/dev/` 目录

# 3 设计蓝图

## 3.1 单实例内部请求流
```
       +-------------------------------------------------------------+
       |                        WEB brower                           |
       +-----------------------------------^-------------------------+
     /                      |              |
    |  +--------------------V--------------|-------------------------+
    |  | +----------------HTTPServer(Threadpool&Queue)-------------+ |
    |  | |   +-------------------+ put +-----------------------+   | |
    |  | |   |ThreadPool(Queue) <------+ HTTPConnection        |   | |
    |  | |   |+---------------+  |     | +-------------------+ |   | |
    |  | |   ||WorkerThread   |  |     | |req=HTTPRequest()  | |   | |
    |  | |   ||+-+ +-+ +-+ +-+|  |     | |req.parse_request()| |   | |（把 socket 字节流，按 HTTP 协议解析）
    |  | |   ||+-+ +-+ +-+ +-+|  |     | |req.respond()#^!^  | |   | |（封装了 WSGIGateway.response)
    |  | |   |+-|---|---|---|-+  |     | +-------------------+ |   | |
    |  | |   +--|---|---|---|----+     +-----------------------+   | |
    |  | +------|---|---|---|-----------------------^--------------+ |
    |  |        |   |   |   |                       |                | WSGI server
    |  | +------V---V---V---V-WSGIGateway(response)-|--------------+ |
    |  | |                       +------------------+-------------+| |
    |  | |+----------------+     | +----------+   +-------------+ || |
    |  | ||   gen environ  |     | |header_set|   |response body| || |
    |  | |+-----+----------+     | +----------+   +-------------+ || |
    |  | |      |                +--^------^----------^-----------+| |
    |  | +------|-------------------|------|----------|------------+ |
    |  +--------|-------------------|------|----------|--------------+
    |  .........|...................|......|..........|......................
    |           |                   |      |          |
    |  +--------V--------+          |      |          |
    |  |       req       |          |      |          |        (1) 封装 environ 为 Request
    |  +-----------------+          |      |          |            生成 reqid
Butterfly       |                   |      |          |
    |           |                   |      |          |      \
    |  +--------V--------+          |      |          |       |(2) 路由
    |  |  apiname_getter |          |      |          |       |    在路由字典中匹配 environ['PATH_INFO']
    |  +-----------------+          |      |          |       |    {
    |           |                   |      |          |       |      '/apidemo/ping':
    |           |                   |      |          |       |        <xlib.protocol_json.Protocol object>,
    |  +--------V--------+ False +--+--+   |          |       |      '/{app}/{handler}':
    |  |is_protocol_exist|------>| 400 |   |          |       |        <xlib.protocol_json.Protocol object>
    |  +-----------------+       +-----+   |          |       |    }
    |           |                          |          |      /
    |           | (protocol_process)       |          |      \
    |           V                          |          |       |(3) 返回 Response
    |  +-----------------+                 |          |       |    参数：第一个参数为 Request 实例化对象 req
    |  | protocol        | Exception    +-----+       |       |          其他参数为 GET 请求参数名
    |  | +-------------+ |------------->| 500 |       |       |    例子：environ['QUERY_STRING']: 'age=16'
    |  | |/app1/handler| |              +-----+       |       |    +-------------handler demo---------------
    |  | |/app2/handler| |Normal+----------------------------+|    |@funcattr.api      # 标识 handler 类型
    |  | +--+-------+--+ |----->|httpstatus, headers, content||    |def demo(req, age):# req + HTTP 请求参数
    |  +----|-------|----+      +----------------------------+|    |   #（状态信息，数据信息，响应头列表）
    |       |       |                                         |    |   return "OK", {"data": age}, []
    |       |       |                                        /     +----------------------------------------
    |       |  +----V----------------------------------+
    |       |  |+---------+  +---------+  +-----------+|
    |       |  ||DiskCache|  |   FSM   |  |APScheduler||       基础公共库
    |       |  |+---------+  +---------+  +-----------+|
    |       |  +---------------------------------------+
    |  +----V------------------------------------------+
    |  |       (Redis ORM) / (MySQL ORM) / RAL         |       数据访问层
     \ +-----------------------------------------------+
```
## 3.2 多实例间消息队列通信

默认实例只能被动接受请求，然后返回响应，通过开启【百川】，实例将主动消费 MQ 中消息

> * 开启方式
>   * Butterfly 服务配置文件中，CACHES 配置开启 `"baichuan"` 的 Redis 连接配置
> * Butterfly 行为
>   * (1) 每分钟发送心跳到 MQ(Redis)
>   * (2) 每 15 分钟进行一次清理 MQ 中过期任务，运行此任务时会加锁
>   * (3) 启动单独线程检查相关的任务，以 `拉` 模式从队列中拉取消息并进行处理，处理结果也会存储在此 Redis 中
> * MQ
>   * (1) Redis 服务，支持 Twemproxy 组成的 Redis 集群
>   * (2) 每个 `/{app}/{handler}` 是个单独的队列
>     * topic: `/{app}/{handler}`
>     * message: 消息为 json（butterfly 消费消息时，会将 json 转为字典以参数传给 handler）
>   * (3) 消息的四种状态 (queued: 还在队列中，started: 执行中，failed: 执行失败，finished: 完成）

```
                    +Message Queue ------------------------------------------+
+---------------+   |                                                        |    +----------------+
|  other  app   |   |                                                        |    |   other app    |
|     ...       |-->| +Canghai Dashboard ---+------+-------+--------+------+ |<---|      ...       |
+---------------+   | |Queue                |Queued|Started|Finished|Failed| |    +----------------+
                    | +---------------------+------+-------+--------+------+ |
                    | |/demo_api/hello      |  0   |   0   |  91    |  0   | |
                    | |/demo_api/ping       |71731 |  50   | 71736  |  27  | |
                    | |/{app}/{handler}     |  X   |   X   |   X    |  X   | |
+ruqi-----------+   | +---------------------+------+-------+--------+------+ |    +baichuan/canghai--+
|port:4100      |   |                                                        |    |port:4300         |
|handlers/ruqi  |-->| +Redis Arch -----------------------------------------+ |<---|handlers/baichuan |
+---------------+   | |               +---------+                          | |    |handlers/canghai  |
                    | |               |Twemproxy| \                        | |    +------------------+
                    | |               +--+---+--+  +----------+            | |
                    | |                  |   |     |Metaserver|            | |
+wuxing---------+   | |               +--V---V--+  +----------+            | |    +xingqiao----------+
|port:4200      |   | |               |  Redis  | /                        | |    |port:4700         |
|handlers/wuxing|-->| |               +---------+                          | |<---|handler/xingqiao  |
+---------------+   | +----------------------------------------------------+ |    +------------------+
                    +--------------------------------------------------------+
```

通用服务能力 (app)：

> * 【如期】而至 --- 定时调度服务
> * 【五行】化一 --- 数据管理服务
> * 【百川】归海 --- 消息队列服务
> * 【星桥】锁月 --- 流程编排服务

butterfly 框架基础能力：

> * 工作流
> * 状态机
> * 定时器
> * 数据库交互 (MySQL/Redis)

# 4 了解更多

## 4.1 传送门

> * 手册
>   * [Butterfly 用户手册](https://github.com/meetbill/butterfly/wiki/usage)
>   * [Butterfly 进阶手册](https://github.com/meetbill/butterfly/wiki)
> * 项目
>   * [Butterfly 示例](https://github.com/meetbill/butterfly-examples)
>   * [Butterfly-fe -- 基于 AngularJS 的前端项目](https://github.com/meetbill/butterfly-fe)
>   * [Butterfly-admin -- 基于 AMIS 的前端项目](https://github.com/meetbill/butterfly-admin)
>   * [Butterfly nginx 配置](https://github.com/meetbill/butterfly-nginx)

## 4.2 自带 app 介绍

看文档不过瘾，还可以通过了解当前 handler 的实现，进而实现自己的需求：

> * 【例子】/handlers/demo_api: 简单 api handler demo
> * 【例子】/handlers/demo_download: 文件下载 handler demo
> * 【例子】/handlers/demo_httpapi: 自定义 HTTP 返回码 handler demo
> * 【例子】/handlers/demo_log: 日志级别调整 handler demo
> * 【例子】/handlers/demo_template: 后端模板 handler demo, 本例子用于输出访问日志统计状态图
> * 【例子】/handlers/demo_async_job: 异步任务 handler demo
> * 【例子】/handlers/demo_stackdump: 打印 Butterfly stack trace 信息到日志
> * 【火眼】/handlers/huoyan: 用于后端接口认证，仅仅抛砖引玉，删除了具体实现
> * 【如期】/handlers/ruqi: 高可用定时任务，可发起 HTTP POST 请求或者执行 Shell/Python 脚本
> * 【五行】/handlers/wuxing: 用于存储配置类 / 监控类 / 巡检类数据，可配合状态机进行使用
> * 【星桥】/handlers/xingqiao: 用于管理 handler 依赖，流程编排

> 备注：
```
可将 examples 下 app 拷贝到 {butterfly_project}/handlers/ 下
```
## 4.3 报告错误

如果您想报告错误，请在 GitHub 上[创建一个新问题](https://github.com/meetbill/butterfly/issues/new)。

# 5 贡献

Use issues for everything

> * For a small change, just send a PR.
> * For bigger changes open an issue for discussion before sending a PR.
> * PR should have:
>   * Test case
>   * Documentation
>   * Example (If it makes sense)
> * You can also contribute by:
>   * Reporting issues
>   * Suggesting new features or enhancements
>   * Improve/fix documentation

# 6 版本信息

本项目的各版本信息和变更历史可以在[这里][changelog] 查看。

# 7 参加步骤

* 在 GitHub 上 `fork` 到自己的仓库，然后 `clone` 到本地，并设置用户信息。
```bash
$ git clone https://github.com/meetbill/butterfly.git
$ cd butterfly
$ git config user.name "yourname"
$ git config user.email "your email"
```
* 修改代码后提交，并推送到自己的仓库。
```bash
$ #do some change on the content
$ git commit -am "Fix issue #1: change helo to hello"
$ git push
```
* 在 GitHub 网站上提交 pull request。
* 定期使用项目仓库内容更新自己仓库内容。
```bash
# 配置原仓库地址
$ git remote add upstream https://github.com/meetbill/butterfly.git
# 查看当前仓库的远程仓库地址和原仓库地址
$ git remote -v
# 获取原仓库的更新，使用 fetch 更新，fetch 后会被存储在一个本地分支 upstream/master 上
$ git fetch upstream
# 切换到本地 master 分支
$ git checkout master
# 合并
$ git rebase upstream/master
# 查看更新
$ git log
# 推送到自己仓库
$ git push -f origin master
# 把本地全部存在的 tag 推送到服务器
$ git push origin --tags
```

[changelog]: CHANGELOG.md


CHANGELOG.md
Changelog
===
以下记录了项目中所有值得关注的变更内容，其格式基于 [Keep a Changelog]。

本项目版本遵守 [Semantic Versioning] 和 [PEP-440]。
## [1.1.14] - 2021-07-20
### Changed
> * 服务:【百川】
>   * 百川 msg 添加 cost 属性，记录消息处理耗时，单位是 s [commit](https://github.com/meetbill/butterfly/commit/58ee024b96d71467819209e081eb0dfe68616387)
>   * 百川标记 failed 策略由 `ERR_` 状态改为仅判断是否为参数错误(客户端错误)和服务端异常(消费者异常) [commit](https://github.com/meetbill/butterfly/commit/0ab7d29b9b406e92549fd85346599b73f9261885)
>   * 百川发布消息时增加多次重试 [commit](https://github.com/meetbill/butterfly/commit/9a2220576e9783c4109e91ef99bfb1c7313fc6fd)

### Added
> * 服务:【星桥】
>   * 增加分布式流程编排功能，有序编排 handler
> * 配置:
>   * 添加 STAT_ADAPTOR 配置，可扩展 API 接口中的状态字段, 以适配前端使用 [commit](https://github.com/meetbill/butterfly/commit/5b852e763b2e32ad35713d8330a66c6c67238f87)
>   * 支持配置 user http header，传对应 header 值为用户名时，会自动记录对应用户操作 [commit](https://github.com/meetbill/butterfly/commit/e7d303756a8d65157ba9329dcdd03b78993920bb)

## [1.1.13] - 2021-06-17
### Changed
> * 服务: 【五行】
>   * 增加 wuxing parameter length check [commit](https://github.com/meetbill/butterfly/commit/1e830331c76b7fb5a2eb7f9a2c18f15dfea124ef)
> * 服务: 【如期】
>   * 定时任务执行日志添加参数信息 [commit](https://github.com/meetbill/butterfly/commit/c082e48b681561a891de9896c3b29fd71ac8bd28)
>   * 去掉 ruqi scheduler_name 配置，改为使用 host_util.server_name 作为 scheduler heduler_name [commit](https://github.com/meetbill/butterfly/commit/2d53fa9f81577d420e1e4aa2785fb4d48f514ea0)
> * 服务:
>   * 将 handlers 目录下 handler 移动至 examples 目录下 [commit](https://github.com/meetbill/butterfly/commit/320e710e01fc35d17760212e27de845b06339a50)

### Fixed
> * 日志
>   *  修复轮转 acc.log 等日志时线程不安全问题 [commit](https://github.com/meetbill/butterfly/commit/b4ac5da3d600faeb6852bdcc6290f4665b94203a)

### Added
> 组件
>   * add pluginbase, 作用: (1)管理 workflow, 通过 name 查找特定 workflow class, (2)小蝶 ChatOps 中通过 name 查找对应 action 函数 [commit](https://github.com/meetbill/butterfly/commit/b334cbbecbdb3fc89b18eb2d9ed5c82d99a1a495)

## [1.1.12] - 2021-05-10
### Changed
> * 服务：【五行】
>   * 适配 item_value 为 bool 时，传值为 int 类型的情况 [commit](https://github.com/meetbill/butterfly/commit/85907a6a572c6acc05d4ff67b83358fd4a79cb28)
>   * 创建 instance 时的 items_data 的 value 传值为 None 时，则使用默认值进行创建 [commit](https://github.com/meetbill/butterfly/commit/14d4bf9f6b4f9f5e362a177b98e8d51bc592f16b)
>   * instance get 返回结果可返回 simple/detail 两种格式，默认是 simple 格式 [commit](https://github.com/meetbill/butterfly/commit/cdaa539feec13baae08b9f45b0911262a513d622)
> * 配置:
>   * 配置中日志和 data 目录从相对路径修改为绝对路径 [commit](https://github.com/meetbill/butterfly/commit/6390d66a5b3fe9418285198612e04a5edf2a4823)
> * 服务：【如期】
>   * ruqi job list 添加 c_time 信息 [commit](https://github.com/meetbill/butterfly/commit/13eca4a5c06271db0fd8ac99fe28c246f30aae0d)
> * 日志
>   * 参数错误时，记录访问的具体参数信息 [commit](https://github.com/meetbill/butterfly/commit/f11c9f801881ceece4cf97f05f23d920e3ce2b8d)
> * RAL
>   * 更新 mcpack 包 [commit](https://github.com/meetbill/butterfly/commit/c2470aed4b35e67c19a3e469e1a600d8140afa45)

### Added
> * 服务: 【五行】
>   * wuxing_cli [commit](https://github.com/meetbill/butterfly/commit/22204633ed7899629738927c8a990c55f3b13955)
> * 服务: 【如期】
>   * 添加 ruqi get_job 接口，用于获取 job 详情 [commit](https://github.com/meetbill/butterfly/commit/163922a3349b24be276f84d946fd13562f3e86d5)
> * 服务：【星轨】
>   * workflow [commit](https://github.com/meetbill/butterfly/commit/fa5024301619e1df8f75d4d1a6671171520e4b81)
> * 调试：
>   * 可将 Butterfly stack trace 信息到日志 [commit](https://github.com/meetbill/butterfly/commit/24f6edd316a889f1d4fc22239d922c8275d965d4)
> * 服务：【如期】
>   * 添加修改定时任务接口 [commit](https://github.com/meetbill/butterfly/commit/fec323c046688f28b7048e844f691082b1e1be4a)

### Fixed
> * 服务：【五行】
>   * 修复创建 item 时，item 类型为 bool 时，没有检查是否为 unicode 类型的问题 [commit](https://github.com/meetbill/butterfly/commit/05418c28de629bf88b55f8e977e24be0fb362cae)
> * 服务：【百川】
>   * 修复消息队列中设置 handle_worker 不生效问题 [commit](https://github.com/meetbill/butterfly/commit/dd4a214471b2a5057c77d8293ab4c30e4eb80cdf)


## [1.1.11] - 2021-03-17
### Changed
- 服务：【五行】(1)baichuan 处理 msg 时改为直接处理 json 数据，而不是进行 urllib.urlencode 数据；(2)baichuan worker 添加 home_dir 属性，用于后期部署管理; (3)baichuan msg 添加 handle_worker 属性，标记此消息的处理 worker

## [1.1.10] - 2021-03-14
### Changed
- 服务：【五行】(1) wuxing item 中增加 text 类型，方便记录长文本数据，此数据没有加索引; (2) item list 增加 c_time 排序
- 服务：【百川】worker 添加 nickname 属性，用于标记此 butterfly 的服务类型, 此配置在 config.SERVER_NAME 中进行配置
- 配置：config.diskcache_dir --> config.LOCALDATA_DIR
- 工具：util.Base64_16 --> util.Base64and16 , 此变动影响 handler 中的 ping 函数

### Added
- 配置：增加 config.SERVER_NAME 配置，目前仅提供给百川 worker 使用

### Other

按照编码规范进行修改部分代码

### Changed
- 服务：【百川】1. 添加 worker 消费限流; 2.worker 每次消费时，将多个队列进行随机打乱再进行消费; 3.handler 处理异常时可以将异常信息进行存储到 mq 中
- 调试：参数支持传 None

## [1.1.9] - 2021-03-05

### Changed
- 服务：【百川】1. 添加 worker 消费限流; 2.worker 每次消费时，将多个队列进行随机打乱再进行消费; 3.handler 处理异常时可以将异常信息进行存储到 mq 中
- 调试：参数支持传 None

### Fixed
- 服务：【五行】handler 库中使用了 instance package, 此目录被 git 忽略, 修改忽略规则

## [1.1.8] - 2021-02-28
### Added

- 服务：添加【百川】app，开启百川后 butterfly 服务将自动消费 MQ 中的消息
- 服务：优化【五行】app，增加五行数据缓存功能

### Changed

- 日志：talk、params、res 的分割符由`,`改为`;`
- 调试：test_handler.py 执行过程中的 logging 日志移动至 logs/dev 目录

## [1.1.7] - 2021-02-11
### Added

- 添加【五行】app，用于实现配置类/监控类数据存储，可配合状态机实现更强大功能
- 添加 demo_async_job, 演示如何发起异步任务

### Changed

- 配置: 支持 MySQL/Redis 多实例支持
- 日志: access 日志字段变更, 更方便进行日志监控
- apscheduler: Merge part of apscheduler-3.7.0 code

## [1.1.6] - 2021-01-09
### Fixed

- 修复解锁已过期任务可能导致的死锁问题 https://github.com/meetbill/butterfly/issues/4
- 修复存在暂停中任务时，调度失败问题 https://github.com/meetbill/butterfly/issues/24

### Added

scheduler manager

- 添加查询 scheduler 状态接口
- 添加强制唤醒 scheduler 接口
- 增加 scheduler http job 方法

### Changed

- 修改 scheduler 部分日志
- http_util 增加 cost 属性

## [1.1.5] - 2021-01-05
### Changed

- update APScheduler 2.1.2 ==> APScheduler 3.6.3
```
(1) add apscheduler mysql jobstores use peewee
(2) 增加 scheduler job_lock，scheduler 检查任务时进行抢锁发起任务
(3) 增加 scheduler manager 作为通用配置
(4) 增加 scheduler 执行历史，存储在 MySQL 中
(5) 增加 scheduler 定期探测是否有任务配置

butterfly/conf/config.py
# coding:utf8
"""
Butterfly config
"""
import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

SERVER_LISTEN_ADDR = ("0.0.0.0", 8585)
SERVER_THREAD_NUM = 16
SERVER_NAME = "Butterfly_app"

# Log
LOG_SIZE_LIMIT = 1024 * 1024 * 2
LOG_BATCH_WRITE = 0
PATH_INIT_LOG = os.path.join(BASE_DIR, "logs/init.log")
PATH_ACC_LOG = os.path.join(BASE_DIR, "logs/acc.log")
PATH_INFO_LOG = os.path.join(BASE_DIR, "logs/info.log")
PATH_WARNING_LOG = os.path.join(BASE_DIR, "logs/warning.log")
PATH_ERR_LOG = os.path.join(BASE_DIR, "logs/err.log")
PATH_CRIT_LOG = os.path.join(BASE_DIR, "logs/crit.log")
PATH_COMMON_LOG = os.path.join(BASE_DIR, "logs/common.log")
PATH_COMMON_BF_LOG = os.path.join(BASE_DIR, "logs/common_bf.log")

# static
STATIC_PATH = "static"
STATIC_PREFIX = "static"

# DB
"""
# wuxing/ruqi use default database
"""
DATABASES = {
    "default": "mysql+retrypool://root:password@127.0.0.1:3306/test?max_connections=300&stale_timeout=300",
}

# Redis
"""
# eg1:"redis://[[username]:[password]]@localhost:6379/0"
# eg2:"redis://@localhost:6379/0?socket_timeout=2&socket_connect_timeout=1&retry_on_timeout=true"
"""
CACHES = {
    "default": "redis://@localhost:6379/0",
    # for cache: connect_timeout 100ms, read_timeout 200ms
    # "wuxing": "redis://@localhost:6379/0?socket_timeout=0.2&socket_connect_timeout=0.1&retry_on_timeout=false",
    # for mq: connect_timeout 500ms, read_timeout 2000ms
    # "baichuan": "redis://@localhost:6379/0?socket_timeout=2&socket_connect_timeout=0.5&retry_on_timeout=false",
}

# Local data or cache
LOCALDATA_DIR = os.path.join(BASE_DIR, "data")

# Auth
SECRET_KEY = None               # If it is None, a key will be randomly generated each time butterfly is started
JWT_TOKEN_TTL = 28800           # default 8 hours
HEADER_USERNAME = "HTTP_X_USERNAME"

# Scheduler
scheduler_store = "memory"      # ("mysql"/"memory")

# Adaptor
"""
# {"stat":"ERR", "data":{}} --> {"stat":"ERR", "success":False, "message":"ERR", "data":{}}
STAT_ADAPTOR={"status_name":"success", "status_map":{"OK":True,"default":False}, "message_name":"message"}
"""
STAT_ADAPTOR = None
